<a href="https://colab.research.google.com/github/1101madan/Association_Rules/blob/main/Association_Rules.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [19]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [20]:
from mlxtend.frequent_patterns import apriori,association_rules
from mlxtend.preprocessing import TransactionEncoder

In [21]:
retail_data = pd.read_excel('/content/drive/MyDrive/Colab Notebooks/Files/Online retail.xlsx',header=None,)

In [22]:
retail_data.head()

0
0  shrimp,almonds,avocado,vegetables mix,green gr...
1                             burgers,meatballs,eggs
2                                            chutney
3                                     turkey,avocado
4  mineral water,milk,energy bar,whole wheat rice...

In [23]:
retail_data.shape

(7501, 1)

In [24]:
retail_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       7501 non-null   object
dtypes: object(1)
memory usage: 58.7+ KB


### Preprocessing

In [25]:
retail_data = retail_data.rename(columns={0:'Products'})

In [26]:
retail_data_3 =  retail_data['Products'].str.split(',').tolist()
retail_data_3[:5]

[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea']]

In [27]:
from mlxtend.preprocessing import TransactionEncoder

In [28]:
encoder = TransactionEncoder().fit(retail_data_3)
onehot = encoder.transform(retail_data_3)
onehot = pd.DataFrame(onehot,columns=encoder.columns_)
onehot.head()

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False     True               True      False     True        False   
1       False    False              False      False    False        False   
2       False    False              False      False    False        False   
3       False    False              False      False     True        False   
4       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False            True   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...   False           False   
3  False           False      False        False  ...    True           False   
4  False           False      False        False  ...   False           False   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0        False       False              True              False   
1        False       False             False              False   
2        False       False             False              False   
3        False       False             False              False   
4        False       False             False              False   

   whole wheat rice   yams  yogurt cake  zucchini  
0             False   True        False     False  
1             False  False        False     False  
2             False  False        False     False  
3             False  False        False     False  
4              True  False        False     False  

[5 rows x 120 columns]

#### Apriori Algorithm

In [29]:
frequent_itemsets_2 = apriori(onehot,min_support=0.009,use_colnames=True)
frequent_itemsets_2

support                                 itemsets
0    0.020397                                (almonds)
1    0.033329                                (avocado)
2    0.010799                         (barbecue sauce)
3    0.014265                              (black tea)
4    0.009199                            (blueberries)
..        ...                                      ...
302  0.017064  (ground beef, spaghetti, mineral water)
303  0.015731         (spaghetti, mineral water, milk)
304  0.010265    (olive oil, spaghetti, mineral water)
305  0.011465     (spaghetti, mineral water, pancakes)
306  0.009332     (spaghetti, mineral water, tomatoes)

[307 rows x 2 columns]

In [30]:
rules2 = association_rules(frequent_itemsets_2,metric='lift',min_threshold=1)
rules2

antecedents                 consequents  \
0                    (avocado)             (mineral water)   
1              (mineral water)                   (avocado)   
2                   (brownies)             (mineral water)   
3              (mineral water)                  (brownies)   
4                       (cake)                   (burgers)   
..                         ...                         ...   
515      (spaghetti, tomatoes)             (mineral water)   
516  (mineral water, tomatoes)                 (spaghetti)   
517                (spaghetti)   (mineral water, tomatoes)   
518            (mineral water)       (spaghetti, tomatoes)   
519                 (tomatoes)  (spaghetti, mineral water)   

     antecedent support  consequent support   support  confidence      lift  \
0              0.033329            0.238368  0.011598    0.348000  1.459926   
1              0.238368            0.033329  0.011598    0.048658  1.459926   
2              0.033729            0.238368  0.009599    0.284585  1.193888   
3              0.238368            0.033729  0.009599    0.040268  1.193888   
4              0.081056            0.087188  0.011465    0.141447  1.622319   
..                  ...                 ...       ...         ...       ...   
515            0.020931            0.238368  0.009332    0.445860  1.870467   
516            0.024397            0.174110  0.009332    0.382514  2.196964   
517            0.174110            0.024397  0.009332    0.053599  2.196964   
518            0.238368            0.020931  0.009332    0.039150  1.870467   
519            0.068391            0.059725  0.009332    0.136452  2.284661   

     leverage  conviction  zhangs_metric  
0    0.003654    1.168147       0.325896  
1    0.003654    1.016113       0.413630  
2    0.001559    1.064601       0.168069  
3    0.001559    1.006814       0.213227  
4    0.004398    1.063198       0.417434  
..        ...         ...            ...  
515  0.004343    1.374439       0.475323  
516  0.005084    1.337503       0.558451  
517  0.005084    1.030856       0.659684  
518  0.004343    1.018962       0.611022  
519  0.005247    1.088851       0.603578  

[520 rows x 10 columns]

In [33]:
# TOP 20 ASSOCIATION RULES
rules2.sort_values(by='lift',ascending=False).head(20)

antecedents                         consequents  \
245                     (herb & pepper)                       (ground beef)   
244                       (ground beef)                     (herb & pepper)   
493                       (ground beef)          (spaghetti, mineral water)   
492          (spaghetti, mineral water)                       (ground beef)   
369                           (burgers)                (french fries, eggs)   
368                (french fries, eggs)                           (burgers)   
486                   (spaghetti, milk)                       (ground beef)   
487                       (ground beef)                   (spaghetti, milk)   
463                       (ground beef)  (frozen vegetables, mineral water)   
462  (frozen vegetables, mineral water)                       (ground beef)   
504          (spaghetti, mineral water)                         (olive oil)   
505                         (olive oil)          (spaghetti, mineral water)   
308                 (whole wheat pasta)                              (milk)   
309                              (milk)                 (whole wheat pasta)   
485                 (ground beef, milk)                         (spaghetti)   
488                         (spaghetti)                 (ground beef, milk)   
215                          (tomatoes)                 (frozen vegetables)   
214                 (frozen vegetables)                          (tomatoes)   
211                            (shrimp)                 (frozen vegetables)   
210                 (frozen vegetables)                            (shrimp)   

     antecedent support  consequent support   support  confidence      lift  \
245            0.049460            0.098254  0.015998    0.323450  3.291994   
244            0.098254            0.049460  0.015998    0.162822  3.291994   
493            0.098254            0.059725  0.017064    0.173677  2.907928   
492            0.059725            0.098254  0.017064    0.285714  2.907928   
369            0.087188            0.036395  0.009065    0.103976  2.856852   
368            0.036395            0.087188  0.009065    0.249084  2.856852   
486            0.035462            0.098254  0.009732    0.274436  2.793141   
487            0.098254            0.035462  0.009732    0.099050  2.793141   
463            0.098254            0.035729  0.009199    0.093623  2.620390   
462            0.035729            0.098254  0.009199    0.257463  2.620390   
504            0.059725            0.065858  0.010265    0.171875  2.609786   
505            0.065858            0.059725  0.010265    0.155870  2.609786   
308            0.029463            0.129583  0.009865    0.334842  2.583999   
309            0.129583            0.029463  0.009865    0.076132  2.583999   
485            0.021997            0.174110  0.009732    0.442424  2.541060   
488            0.174110            0.021997  0.009732    0.055896  2.541060   
215            0.068391            0.095321  0.016131    0.235867  2.474464   
214            0.095321            0.068391  0.016131    0.169231  2.474464   
211            0.071457            0.095321  0.016664    0.233209  2.446574   
210            0.095321            0.071457  0.016664    0.174825  2.446574   

     leverage  conviction  zhangs_metric  
245  0.011138    1.332860       0.732460  
244  0.011138    1.135410       0.772094  
493  0.011196    1.137902       0.727602  
492  0.011196    1.262445       0.697788  
369  0.005892    1.075422       0.712046  
368  0.005892    1.215598       0.674513  
486  0.006248    1.242822       0.665583  
487  0.006248    1.070579       0.711930  
463  0.005688    1.063874       0.685755  
462  0.005688    1.214412       0.641290  
504  0.006332    1.128021       0.656007  
505  0.006332    1.113898       0.660314  
308  0.006047    1.308587       0.631612  
309  0.006047    1.050515       0.704263  
485  0.005902    1.481216       0.620104  
488  0.005902    1.0

> ### As in the Association Rules above we can see that the herb & Pepper and ground beef have highest association. (Lift - 3.29)
> ### It may be because Ground beef is a staple diet and is cooked with herb & Pepper often.

> ### Based on the Association rules herb & pepper, spaghetti, mineral water, milk, frozen vegetables can be placed near Ground beef.

> ### Burger, French fries and eggs are often purchased together
> ### It is a popular combination of fast food.

> ### Customers are more likely to buy Products like milk, ground beef, olive oil, mineral water along with Spaghetti.

> What is Lift and Why is it important in Association Rules?
* Lift is the Measure of how strong is the Association betweeen antecedant and consequent.
* It is a measure which takes into the account bias that is introduced in the association purly by chance.

> What is Support and Confidence? How do we calculate them?
* Support is the proportion of transactions in the dataset that contain both the antecedent and the consequent..
* Support =
Number of transaction with Antecedent and consequent / Total Transactions
* Confidence measures the likelihood that the consequent will be purchased when the antecedent is purchased.
* Confidence = Number of transaction with Antecedent and consequent / total andecedent
* Can also be given by :
  *Support = AnB/N
  *Confidence = AnB/A

> What are some limitations or challanges of association rules mining?
* Association rules mining are not best when the dataset is large with multiple features
* Association mining can be carried out only for categorical data.
* Determining the threshold value for minimum support value in apriori algorithm is only by trial and error.
* Sometimes the association that are found can be trivial or be inexplicable. Idnetifying the actionable Rules is important out of all the rules



